java/configuration/src/main/java/sleeper/configuration/properties/table/TablePropertiesProvider.java

```diff
@@ -16,28 +16,54 @@
 package sleeper.configuration.properties.table;
 
 import com.amazonaws.services.s3.AmazonS3;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import sleeper.configuration.properties.instance.InstanceProperties;
 
+import java.time.Duration;
+import java.time.Instant;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.Optional;
 import java.util.function.Function;
+import java.util.function.Supplier;
+
+import static sleeper.configuration.properties.instance.CommonProperty.TABLE_PROPERTIES_PROVIDER_TIMEOUT_IN_MINS;
 
 public class TablePropertiesProvider {
+    private static final Logger LOGGER = LoggerFactory.getLogger(TablePropertiesProvider.class);
     private final Map<String, TableProperties> tableNameToPropertiesCache;
     private final Function<String, TableProperties> getTableProperties;
+    private Supplier<Instant> timeSupplier;
+    private final int timeoutInMins;
+    private Instant expireTime;
 
     public TablePropertiesProvider(AmazonS3 s3Client, InstanceProperties instanceProperties) {
-        this(tableName -> getTablePropertiesFromS3(s3Client, instanceProperties, tableName));
+        this(s3Client, instanceProperties, Instant::now);
+    }
+
+    protected TablePropertiesProvider(AmazonS3 s3Client, InstanceProperties instanceProperties, Supplier<Instant> timeSupplier) {
+        this(tableName -> getTablePropertiesFromS3(s3Client, instanceProperties, tableName),
+                instanceProperties.getInt(TABLE_PROPERTIES_PROVIDER_TIMEOUT_IN_MINS), timeSupplier);
     }
 
-    protected TablePropertiesProvider(Function<String, TableProperties> getTableProperties) {
+    protected TablePropertiesProvider(Function<String, TableProperties> getTableProperties,
+                                      int timeoutInMins, Supplier<Instant> timeSupplier) {
         this.getTableProperties = getTableProperties;
         this.tableNameToPropertiesCache = new HashMap<>();
+        this.timeoutInMins = timeoutInMins;
+        this.timeSupplier = timeSupplier;
+        this.expireTime = timeSupplier.get().plus(Duration.ofMinutes(timeoutInMins));
     }
 
     public TableProperties getTableProperties(String tableName) {
+        Instant currentTime = timeSupplier.get();
+        if (currentTime.isAfter(expireTime)) {
+            LOGGER.info("Table properties provider expiry time reached, clearing cache.");
+            clearCache();
```

Comment:
I think we want to reload the properties for a table if it's been more than N minutes since the properties for that table were loaded, not refresh the entire cache after N minutes. With refreshing the entire cache after N minutes, we could load the properties for a table N-1 minutes after the cache was last cleared and then reload the properties for that table after 1 minute.